File: server/routes/api/habitProgress.js
const express = require('express');
const router = express.Router();
const auth = require('../../middleware/auth');
const { check, validationResult } = require('express-validator');
const { Op } = require('sequelize');

// Models
const Habit = require('../../models/Habit');
const HabitProgress = require('../../models/HabitProgress');

// @route   GET api/habit-progress
// @desc    Get all habit progress entries for the authenticated user
// @access  Private
router.get('/', auth, async (req, res) => {
  try {
    const habitProgress = await HabitProgress.findAll({
      include: [{
        model: Habit,
        where: { userId: req.user.id },
        attributes: ['name', 'type', 'target', 'unit']
      }],
      order: [['date', 'DESC']]
    });

    res.json(habitProgress);
  } catch (err) {
    console.error(err.message);
    res.status(500).send('Server Error');
  }
});

// @route   GET api/habit-progress/by-date
// @desc    Get habit progress for a specific date range
// @access  Private
router.get('/by-date', auth, async (req, res) => {
  try {
    const { startDate, endDate } = req.query;
    
    if (!startDate || !endDate) {
      return res.status(400).json({ msg: 'Start date and end date are required' });
    }

    const habitProgress = await HabitProgress.findAll({
      include: [{
        model: Habit,
        where: { userId: req.user.id },
        attributes: ['name', 'type', 'target', 'unit']
      }],
      where: {
        date: {
          [Op.between]: [new Date(startDate), new Date(endDate)]
        }
      },
      order: [['date', 'ASC']]
    });

    res.json(habitProgress);
  } catch (err) {
    console.error(err.message);
    res.status(500).send('Server Error');
  }
});

// @route   GET api/habit-progress/:habitId
// @desc    Get all progress for a specific habit
// @access  Private
router.get('/:habitId', auth, async (req, res) => {
  try {
    // Verify the habit belongs to the user
    const habit = await Habit.findOne({
      where: {
        id: req.params.habitId,
        userId: req.user.id
      }
    });

    if (!habit) {
      return res.status(404).json({ msg: 'Habit not found or not authorized' });
    }

    const habitProgress = await HabitProgress.findAll({
      where: { habitId: req.params.habitId },
      order: [['date', 'DESC']]
    });

    res.json(habitProgress);
  } catch (err) {
    console.error(err.message);
    res.status(500).send('Server Error');
  }
});

// @route   POST api/habit-progress/:habitId
// @desc    Add or update progress for a habit
// @access  Private
router.post(
  '/:habitId',
  [
    auth,
    [
      check('date', 'Date is required').not().isEmpty(),
      check('value', 'Progress value is required').not().isEmpty(),
      check('notes', 'Notes must be a string').optional().isString()
    ]
  ],
  async (req, res) => {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const { date, value, notes } = req.body;

    try {
      // Verify the habit belongs to the user
      const habit = await Habit.findOne({
        where: {
          id: req.params.habitId,
          userId: req.user.id
        }
      });

      if (!habit) {
        return res.status(404).json({ msg: 'Habit not found or not authorized' });
      }

      // Check if there's already progress for this date
      let habitProgress = await HabitProgress.findOne({
        where: {
          habitId: req.params.habitId,
          date: new Date(date)
        }
      });

      if (habitProgress) {
        // Update existing progress
        habitProgress.value = value;
        if (notes) habitProgress.notes = notes;
        await habitProgress.save();
      } else {
        // Create new progress entry
        habitProgress = await HabitProgress.create({
          habitId: req.params.habitId,
          date: new Date(date),
          value,
          notes: notes || ''
        });
      }

      res.json(habitProgress);
    } catch (err) {
      console.error(err.message);
      res.status(500).send('Server Error');
    }
  }
);

// @route   DELETE api/habit-progress/:id
// @desc    Delete a habit progress entry
// @access  Private
router.delete('/:id', auth, async (req, res) => {
  try {
    const habitProgress = await HabitProgress.findByPk(req.params.id);

    if (!habitProgress) {
      return res.status(404).json({ msg: 'Progress entry not found' });
    }

    // Verify the habit belongs to the user
    const habit = await Habit.findOne({
      where: {
        id: habitProgress.habitId,
        userId: req.user.id
      }
    });

    if (!habit) {
      return res.status(401).json({ msg: 'Not authorized to delete this progress entry' });
    }

    await habitProgress.destroy();
    res.json({ msg: 'Progress entry removed' });
  } catch (err) {
    console.error(err.message);
    res.status(500).send('Server Error');
  }
});

module.exports = router; 

File: flask-backend/routes/habit_progress.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from werkzeug.exceptions import BadRequest, NotFound, Forbidden
from sqlalchemy import and_, between
from models import Habit, HabitProgress, db
from datetime import datetime, timedelta

habit_progress_bp = Blueprint('habit_progress', __name__)

@habit_progress_bp.route('/<int:habit_id>', methods=['GET'])
@jwt_required()
def get_habit_progress(habit_id):
    """Get progress for a specific habit"""
    try:
        user_id = get_jwt_identity()
        
        # Check if habit exists and belongs to user
        habit = Habit.query.get(habit_id)
        if not habit:
            raise NotFound('Habit not found')
        if habit.userId != user_id:
            raise Forbidden('Not authorized to access this habit')
        
        # Get date range from query parameters
        start_date_str = request.args.get('startDate')
        end_date_str = request.args.get('endDate')
        
        # Default to last 30 days if no dates provided
        end_date = datetime.utcnow().date()
        start_date = end_date - timedelta(days=30)
        
        if start_date_str:
            start_date = datetime.fromisoformat(start_date_str.replace('Z', '+00:00')).date()
        if end_date_str:
            end_date = datetime.fromisoformat(end_date_str.replace('Z', '+00:00')).date()
        
        # Query progress records within date range
        progress = HabitProgress.query.filter(
            and_(
                HabitProgress.habitId == habit_id,
                HabitProgress.date.between(start_date, end_date)
            )
        ).order_by(HabitProgress.date).all()
        
        return jsonify([p.to_dict() for p in progress])
    except (NotFound, Forbidden) as e:
        return jsonify({'error': str(e)}), e.code
    except Exception as e:
        return jsonify({'error': str(e) or 'Server error'}), 500

@habit_progress_bp.route('/<int:habit_id>', methods=['POST'])
@jwt_required()
def add_habit_progress(habit_id):
    """Add or update progress for a habit"""
    try:
        user_id = get_jwt_identity()
        
        # Check if habit exists and belongs to user
        habit = Habit.query.get(habit_id)
        if not habit:
            raise NotFound('Habit not found')
        if habit.userId != user_id:
            raise Forbidden('Not authorized to update this habit')
        
        data = request.json
        if not data:
            raise BadRequest('Request body is required')
        
        # Get date from request or default to today
        date_str = data.get('date')
        progress_date = datetime.utcnow().date()
        if date_str:
            progress_date = datetime.fromisoformat(date_str.replace('Z', '+00:00')).date()
        
        # Check if progress already exists for this date
        existing_progress = HabitProgress.query.filter(
            and_(
                HabitProgress.habitId == habit_id,
                HabitProgress.date == progress_date
            )
        ).first()
        
        if existing_progress:
            # Update existing progress
            existing_progress.value = data.get('value', existing_progress.value)
            existing_progress.notes = data.get('notes', existing_progress.notes)
            db.session.commit()
            return jsonify(existing_progress.to_dict())
        else:
            # Create new progress entry
            progress = HabitProgress(
                habitId=habit_id,
                date=progress_date,
                value=data.get('value', 0),
                notes=data.get('notes')
            )
            
            db.session.add(progress)
            db.session.commit()
            return jsonify(progress.to_dict()), 201
    except (BadRequest, NotFound, Forbidden) as e:
        return jsonify({'error': str(e)}), e.code
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e) or 'Server error'}), 500

@habit_progress_bp.route('/<int:progress_id>/delete', methods=['DELETE'])
@jwt_required()
def delete_habit_progress(progress_id):
    """Delete a habit progress entry"""
    try:
        user_id = get_jwt_identity()
        
        # Get the progress record
        progress = HabitProgress.query.get(progress_id)
        if not progress:
            raise NotFound('Progress record not found')
        
        # Check if the associated habit belongs to the user
        habit = Habit.query.get(progress.habitId)
        if not habit or habit.userId != user_id:
            raise Forbidden('Not authorized to delete this progress record')
        
        db.session.delete(progress)
        db.session.commit()
        
        return jsonify({'message': 'Progress record deleted successfully'})
    except (NotFound, Forbidden) as e:
        return jsonify({'error': str(e)}), e.code
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e) or 'Server error'}), 500 